python/lsst/daf/butler/core/_butlerUri.py

```diff
@@ -0,0 +1,1332 @@
+
+    def as_local(self) -> Tuple[str, bool]:
+        """Download object over HTTP and place in temporary directory.
+
+        Returns
+        -------
+        path : `str`
+            Path to local temporary file.
+        temporary : `bool`
+            Always returns `True`. This is always a temporary file.
+        """
+        r = requests.get(self.geturl(), stream=True)
+        if r.status_code != 200:
+            raise FileNotFoundError(f"Unable to download resource {self}; status code: {r.status_code}")
+        with tempfile.NamedTemporaryFile(suffix=self.getExtension(), delete=False) as tmpFile:
+            for chunk in r.iter_content():
+                tmpFile.write(chunk)
+        return tmpFile.name, True
+
+    def read(self, size: int = -1) -> bytes:
+        # Docstring inherits
+        stream = True if size > 0 else False
+        r = requests.get(self.geturl(), stream=stream)
+        if not stream:
+            return r.content
+        else:
+            return next(r.iter_content(chunk_size=size))
+
+
+class ButlerGenericURI(ButlerURI):
+    """Generic URI with a defined scheme"""
+
+    def exists(self) -> bool:
+        """Test for existence and always return False."""
+        return False
+
+    def as_local(self) -> Tuple[str, bool]:
+        raise RuntimeError(f"Do not know how to retrieve data for URI '{self}'")
+
+
+class ButlerSchemelessURI(ButlerFileURI):
+    """Scheme-less URI referring to the local file system"""
+
+    _pathLib = PurePath
+    _pathModule = os.path
+    quotePaths = False
+
+    @property
+    def ospath(self) -> str:
+        """Path component of the URI localized to current OS."""
+        return self.path
+
+    def isabs(self) -> bool:
+        """Indicate that the resource is fully specified.
+
+        For non-schemeless URIs this is always true.
+
+        Returns
+        -------
+        isabs : `bool`
+            `True` if the file is absolute, `False` otherwise.
+        """
+        return os.path.isabs(self.ospath)
+
+    def _force_to_file(self) -> ButlerFileURI:
+        """Force a schemeless URI to a file URI and returns a new URI.
+
+        This will include URI quoting of the path.
+
+        Returns
+        -------
+        file : `ButlerFileURI`
+            A copy of the URI using file scheme. If already a file scheme
+            the copy will be identical.
+
+        Raises
+        ------
+        ValueError
+            Raised if this URI is schemeless and relative path and so can
+            not be forced to file absolute path without context.
+        """
+        if not self.isabs():
+            raise RuntimeError(f"Internal error: Can not force {self} to absolute file URI")
+        uri = self._uri._replace(scheme="file", path=urllib.parse.quote(os2posix(self.path)))
+        # mypy really wants a ButlerFileURI to be returned here
+        return ButlerURI(uri, forceDirectory=self.dirLike)  # type: ignore
```

Comment:
You know that the return of `ButlerURI` will be a `ButlerFileURI`, so rather than `# type: ignore` I would add an explicit `cast`.

Reply:
Thanks. That worked.